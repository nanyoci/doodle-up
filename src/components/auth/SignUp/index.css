#splat-blue-corner {
    position: absolute;
    display: block;
    bottom: 0px;
    right: 0px;
    width: 500px;
    transform: rotate(90deg);
    overflow: hidden;
}
