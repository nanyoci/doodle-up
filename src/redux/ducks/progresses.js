import { STATUSES, METHODS, createApiAction } from './helpers';

export const ENTITY_NAME = 'progresses';

// REDUCER
const initialState = {
    loading: {
        [METHODS.RETRIEVE]: true,
        [METHODS.UPDATE]: false,
        [METHODS.DELETE]: false,
        [METHODS.LIST]: true,
    },
    error: {
        [METHODS.RETRIEVE]: null,
        [METHODS.UPDATE]: null,
        [METHODS.DELETE]: null,
        [METHODS.LIST]: null,
    },
    items: [],
    item: null,
    itemId: null,
};

function progressesReducer(state = initialState, action) {
    const actionTypePattern = /^(.+)_(.+)_(.+)$/;
    const match = action.type.match(actionTypePattern);

    if (!match)
        return state;

    const actionEntityName = match[1];
    const actionStatus = match[2];
    const actionMethod = match[3];

    if (actionEntityName !== ENTITY_NAME.toUpperCase() || !Object.values(METHODS).includes(actionMethod))
        return state;

    switch (actionStatus) {
        case STATUSES.REQUEST:
            return {
                ...state,
                loading: {
                    ...state.loading,
                    [actionMethod]: true
                },
                error: {
                    ...state.error,
                    [actionMethod]: null
                },
                itemId: (action.payload) ? action.payload : state.itemId,
            };

        case STATUSES.SUCCESS:
            switch (actionMethod) {
                case METHODS.RETRIEVE:
                    return {
                        ...state,
                        loading: {
                            ...state.loading,
                            [actionMethod]: false
                        },
                        item: action.payload,
                    }
                
                case METHODS.UPDATE:
                    return {
                        ...state,
                        loading: {
                            ...state.loading,
                            [actionMethod]: false
                        },
                        item: {
                            stages: [...state.item.stages.filter(stageProgress => stageProgress.stage_id !== action.payload.stage_id), action.payload]
                        },
                    }
                
                case METHODS.DELETE:
                    return {
                        ...state,
                        loading: {
                            ...state.loading,
                            [actionMethod]: false,
                        },
                        item: {
                            completed: false,
                            stages: [],
                        },
                    }

                case METHODS.LIST:
                    return {
                        ...state,
                        loading: {
                            ...state.loading,
                            [actionMethod]: false
                        },
                        items: action.payload,
                    }

                default:
                    return state;
            }

        case STATUSES.FAILURE:
            return {
                ...state,
                loading: {
                    ...state.loading,
                    [actionMethod]: false
                },
                error: {
                    ...state.error,
                    [actionMethod]: action.payload
                },
            };

        default:
            return state;
    }

}
export default progressesReducer;

// OPERATIONS
export const updateProgressStage = (storyId, stageId, drawing=null) => (dispatch) => {
    dispatch(createApiAction(ENTITY_NAME, STATUSES.REQUEST, METHODS.UPDATE), storyId);
    dispatch(createApiAction(ENTITY_NAME, STATUSES.SUCCESS, METHODS.UPDATE, {
        stage_id: stageId,
        completed: true,
        image_url: drawing,
    }));
};

export const retrieveProgress = (storyId) => (dispatch) => {
    dispatch(createApiAction(ENTITY_NAME, STATUSES.REQUEST, METHODS.RETRIEVE, storyId));
    dispatch(createApiAction(ENTITY_NAME, STATUSES.SUCCESS, METHODS.RETRIEVE, {
        stages: [],
        id: storyId,
    }));
};

// SELECTORS
export const selectProgressLoading = state => state.progressesReducer.loading[METHODS.RETRIEVE] === true;
export const selectProgressError = state => state.progressesReducer.error[METHODS.RETRIEVE];
export const selectProgress = state => state.progressesReducer.item;

export const selectProgressUpdateLoading = state => state.progressesReducer.loading[METHODS.UPDATE] === true;
