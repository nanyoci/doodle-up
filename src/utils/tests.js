import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { render } from '@testing-library/react';

import rootReducer from '../redux/rootReducer';

export const renderWithRedux = (component, initialState = {}) => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )

    const renderedComponent = <Provider store={store}>{component}</Provider>;

    return {
        ...render(renderedComponent),
        store
    };
}

export const renderWithRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    const renderedComponent = <Router history={history}>{component}</Router>;

    return {
        ...render(renderedComponent),
        history
    };
}

export const renderWithReduxRouter = (component, initialState = {}, route = '/') => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )
    const history = createMemoryHistory({ initialEntries: [route] });

    const renderedComponent = (
        <Provider store={store}>
            <Router history={history}>{component}</Router>
        </Provider>
    );

    return {
        ...render(renderedComponent),
        history,
        store
    };
}