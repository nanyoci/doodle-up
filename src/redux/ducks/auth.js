import axios from 'axios';
// import Cookies from 'js-cookie';

import { API_URL } from '../../utils/constants';
// import { getActionTypes, createAction } from './helpers';
// import { displayError } from './errors';

// ACTION TYPES
export const REGISTER = 'REGISTER';
export const LOGIN = 'LOGIN';
export const LOGOUT = 'LOGOUT';

const initialState = {
	// userLoading: true,
	// userFailed: null,
	// user: {},
	username: localStorage.getItem("username")
};
export default function (state = initialState, action) {
	switch (action.type) {
		case LOGIN:

			// const currentTime = getCurrentTime()

			localStorage.setItem("username", action.payload);

			return {
				...state,
				username: action.payload
			}

		case LOGOUT:
			localStorage.removeItem("username");

			return {
				...state,
				// user: {},
				username: ''
			}

		default:
			return state;
	}

}

// ACTION CREATORS
export function registerAction(payload) {
	return {
		type: REGISTER,
		payload: payload,
	}
}
export function loginAction(payload) {
	return {
		type: LOGIN,
		payload: payload,
	}
}

export function logoutAction() {
	return {
		type: LOGOUT,
	}
}

// OPERATIONS
export const authenticateLogin = userData => dispatch => {
	var formdata = new FormData();
	formdata.append("email", userData.email);
	formdata.append("password", userData.password);
	console.log("authenticating login: ", userData)
	axios
		.post(
			`${API_URL}/signin`,
			formdata,
			{
				'Content-Type': 'application/x-www-form-urlencoded',
				"Access-Control-Allow-Origin": "*",
			},
		)
		.then(res => {
			console.log(res.data)
			dispatch(loginAction(res.data));
		})
		.catch(err => {
			alert("Failed to login")
			// displayError("Unable to login")(dispatch);
			// dispatch(fetchMeFailureAction());
		});
};

export const register = userData => dispatch => {
	var formdata = new FormData();
	formdata.append("email", userData.email);
	formdata.append("username", userData.username);
	formdata.append("password", userData.password);

	axios
		.post(
			`${API_URL}/signup`,
			// `${API_URL}/register`,
			formdata,
		)
		.then(res => {
			if (res.data == "User created") {
				dispatch(registerAction(userData.username))
			}
			else {
				if (res.data == "The email is already in use") {
					alert("The email is already taken. Please try to sign in")
				}
				else {
					alert("200")
				}

			}
		})
		.catch(err => {
			alert("Failed to login")
			// displayError("Unable to login")(dispatch);
			// dispatch(fetchMeFailureAction());
		});
};


export const logout = () => (dispatch, getState) => {
	// No endpoints for logout
	dispatch(logoutAction());
	// axios
	//     .delete(
	//         `${API_URL}/oauth/revoke`,
	//         getTokenConfig(getState)
	//     )
	//     .then(() => {
	//         dispatch(logoutAction());
	//     })
	//     .catch(err => {
	//         displayError("Unable to logout")(dispatch);
	//     });
};

// export const refreshTokenLogin = () => (dispatch, getState) => {
//     var formdata = new FormData();
//     formdata.append("refresh_token", getState().authReducer.refresh_token);
//     formdata.append("grant_type", "refresh_token");

//     axios
//         .post(
//             `${API_URL}/oauth/token`,
//             formdata,
//             {
//                 headers: {
//                     'Authorization': `Basic ${btoa('my-client:my-secret')}`
//                 }
//             },
//         )
//         .then(res => {
//             fetchMe(res.data.access_token)(dispatch);
//             dispatch(loginAction(res.data));
//         })
//         .catch(err => {
//             displayError("Unable to login")(dispatch);
//             dispatch(fetchMeFailureAction());
//         });
// };

// SELECTOR
export const selectUsername = state => state.authReducer.username;
