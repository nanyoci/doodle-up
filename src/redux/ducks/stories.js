import axios from 'axios';

import { API_URL } from '../../utils/constants';
import { STATUSES, METHODS, createApiAction, createApiReducer, displayErrorMsgOrUnauth } from './helpers';

export const ENTITY_NAME = 'stories';

// REDUCER
const storiesReducer = createApiReducer(ENTITY_NAME);
export default storiesReducer;

// OPERATIONS
export const retrieveStory = (storyId) => (dispatch) => {
	dispatch(createApiAction(ENTITY_NAME, STATUSES.REQUEST, METHODS.RETRIEVE, storyId));

	return axios.get(
		`${API_URL}/content`, {
			params: {
				storyid: storyId
			}
		}
	)
		.then(res => {
			console.log(res.data)
			dispatch(createApiAction(ENTITY_NAME, STATUSES.SUCCESS, METHODS.RETRIEVE, res.data));
		})
		.catch(err => {
			displayErrorMsgOrUnauth(err, dispatch, "Unable to fetch story.");
			dispatch(createApiAction(ENTITY_NAME, STATUSES.FAILURE, METHODS.RETRIEVE, err));
		});
	;
};

export const listStories = () => (dispatch) => {
	dispatch(createApiAction(ENTITY_NAME, STATUSES.REQUEST, METHODS.LIST));

	return axios.get(
		`${API_URL}/content/all`,
	)
		.then(res => {
			dispatch(createApiAction(ENTITY_NAME, STATUSES.SUCCESS, METHODS.LIST, res.data.results));
			console.log(res.data.results)
		})
		.catch(err => {
			displayErrorMsgOrUnauth(err, dispatch, "Unable to fetch stories.");
			dispatch(createApiAction(ENTITY_NAME, STATUSES.FAILURE, METHODS.LIST, err));
		});
};

// SELECTORS
export const selectStoriesLoading = state => state.storiesReducer.loading[METHODS.LIST] === true;
export const selectStoriesError = state => state.storiesReducer.error[METHODS.LIST];
export const selectStories = state => state.storiesReducer.items;

export const selectStoryLoading = state => state.storiesReducer.loading[METHODS.RETRIEVE] === true;
export const selectStoryError = state => state.storiesReducer.error[METHODS.RETRIEVE];
export const selectStory = state => state.storiesReducer.item;
export const selectStoryId = state => state.storiesReducer.itemId;
