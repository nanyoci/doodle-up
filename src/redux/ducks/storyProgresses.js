import axios from 'axios';

import { API_URL } from '../../utils/constants';
import { STATUSES, METHODS, createApiAction, createApiReducer, getTokenConfig, displayErrorMsgOrUnauth } from './helpers';

export const ENTITY_NAME = 'storyprogresses';

// REDUCER
const storyProgressesReducer = createApiReducer(ENTITY_NAME);
export default storyProgressesReducer;

// OPERATIONS
export const createStoryProgress = (storyId, { event_type, date, description }) => (dispatch, getState) => {
    dispatch(createApiAction(ENTITY_NAME, STATUSES.REQUEST, METHODS.CREATE));

    return axios.post(
        `${API_URL}/stories/${storyId}/progress/`,
        {
            event_type,
            date,
            description,
        },
        getTokenConfig(getState)
    )
        .then(res => {
            dispatch(createApiAction(ENTITY_NAME, STATUSES.SUCCESS, METHODS.CREATE, res.data));
        })
        .catch(err => {
            displayErrorMsgOrUnauth(err, dispatch, "Unable to create progress.");
            dispatch(createApiAction(ENTITY_NAME, STATUSES.FAILURE, METHODS.CREATE, err));
        });
    ;
};

export const retrieveStoryProgress = (storyId, storyProgressId) => (dispatch, getState) => {
    dispatch(createApiAction(ENTITY_NAME, STATUSES.REQUEST, METHODS.RETRIEVE, storyProgressId));

    return axios.get(
        `${API_URL}/stories/${storyId}/progress/${storyProgressId}/`,
        getTokenConfig(getState)
    )
        .then(res => {
            dispatch(createApiAction(ENTITY_NAME, STATUSES.SUCCESS, METHODS.RETRIEVE, res.data));
        })
        .catch(err => {
            displayErrorMsgOrUnauth(err, dispatch, "Unable to fetch progress.");
            dispatch(createApiAction(ENTITY_NAME, STATUSES.FAILURE, METHODS.RETRIEVE, err));
        });
    ;
};

export const updateStoryProgress = (storyId, { id, event_type, date, description }) => (dispatch, getState) => {
    dispatch(createApiAction(ENTITY_NAME, STATUSES.REQUEST, METHODS.UPDATE, id));

    return axios.patch(
        `${API_URL}/stories/${storyId}/progress/${id}/`,
        {
            event_type,
            date,
            description,
        },
        getTokenConfig(getState)
    )
        .then(res => {
            dispatch(createApiAction(ENTITY_NAME, STATUSES.SUCCESS, METHODS.UPDATE, res.data));
        })
        .catch(err => {
            displayErrorMsgOrUnauth(err, dispatch, "Unable to update progress.");
            dispatch(createApiAction(ENTITY_NAME, STATUSES.FAILURE, METHODS.UPDATE, err));
        });
};

export const deleteStoryProgress = (storyId, storyProgressId) => (dispatch, getState) => {
    dispatch(createApiAction(ENTITY_NAME, STATUSES.REQUEST, METHODS.DELETE));

    return axios.delete(
        `${API_URL}/stories/${storyId}/progress/${storyProgressId}/`,
        getTokenConfig(getState)
    )
        .then(() => {
            dispatch(createApiAction(ENTITY_NAME, STATUSES.SUCCESS, METHODS.DELETE, storyProgressId));
        })
        .catch(err => {
            displayErrorMsgOrUnauth(err, dispatch, "Unable to delete progress.");
            dispatch(createApiAction(ENTITY_NAME, STATUSES.FAILURE, METHODS.DELETE, err));
        });
};

export const listStoryProgresses = storyId => (dispatch, getState) => {
    dispatch(createApiAction(ENTITY_NAME, STATUSES.REQUEST, METHODS.LIST));

    return axios.get(
        `${API_URL}/stories/${storyId}/progress/`,
        getTokenConfig(getState),
    )
        .then(res => {
            dispatch(createApiAction(ENTITY_NAME, STATUSES.SUCCESS, METHODS.LIST, res.data));
        })
        .catch(err => {
            displayErrorMsgOrUnauth(err, dispatch, "Unable to fetch progresses.");
            dispatch(createApiAction(ENTITY_NAME, STATUSES.FAILURE, METHODS.LIST, err));
        });
};

// SELECTORS
export const selectStoryProgressesLoading = state => state.storyProgressesReducer.loading[METHODS.LIST] === true;
export const selectStoryProgressesError = state => state.storyProgressesReducer.error[METHODS.LIST];
export const selectStoryProgresses = state => state.storyProgressesReducer.items;

export const selectStoryProgressLoading = state => state.storyProgressesReducer.loading[METHODS.RETRIEVE] === true;
export const selectStoryProgressError = state => state.storyProgressesReducer.error[METHODS.RETRIEVE];
export const selectStoryProgress = state => state.storyProgressesReducer.item;
