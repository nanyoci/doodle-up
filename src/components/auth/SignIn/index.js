import React, { useState, useEffect } from 'react';
import useSound from 'use-sound';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { authenticateLogin, selectAuthError, selectAuthLoading, authInit } from '../../../redux/ducks/auth';


import './index.css';

import Page from '../../common/Page';
import buttonClickSound from '../../../assets/soundFX/buttonClick.mp3';

function SignInPage({ authenticateLogin, error, isLoading, authInit }) {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  useEffect(() => {
    authInit()
  }, [authInit]);

  const [playMouseClickSound] = useSound(
    buttonClickSound,
    {
      volume: 0.5
    }
  )
  const onSubmit = e => {
    e.preventDefault();
    playMouseClickSound()
    authenticateLogin({ email, password })
  }

  return (
    <Page isMain={true} isNotLoggedIn={true} isLoading={isLoading} >
      <div className="auth-container">
        <h1>Sign In</h1>
        <form className="auth-form" onSubmit={onSubmit}>
          <div className="fill-in-form">
            <div className="fill-in">
              <h2 className="fill-name">Email</h2>
              <input
                data-testid="emailField"
                className="input"
                name="email"
                value={email}
                onChange={e => setEmail(e.target.value)}
                type="text"
                placeholder="Email Address"
              />
            </div>
            <div className="fill-in">
              <h2 className="fill-name">Password</h2>
              <input
                data-testid="passwordField"
                className="input"
                name="password"
                value={password}
                onChange={e => setPassword(e.target.value)}
                type="password"
                placeholder="Password"
              />
            </div>
            <div className="menu-buttons">
              {error &&
                <div>
                  <p data-testid="auth-error-signin" className="auth-error">{error}</p>
                  <br />
                </div>}
            </div>
            <div className=" menu-buttons">
              <button data-testid="loginButton" id="loginButton" type="submit" className="btn btn-primary">Sign In</button>
            </div>
          </div>
        </form>
        <br />
        <Link className="auth-link" to="signup">Create an account? Sign up</Link>
      </div>
    </Page>
  );

}

const mapStateToProps = state => ({
  error: selectAuthError(state),
  isLoading: selectAuthLoading(state),
});

const dispatchers = {
  authenticateLogin,
  authInit
};

export default connect(mapStateToProps, dispatchers)(SignInPage);
