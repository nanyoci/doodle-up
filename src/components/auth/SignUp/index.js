import React, { useState, useEffect } from 'react';
import useSound from 'use-sound';
import { connect } from 'react-redux';
import { authenticateSignUp, selectAuthError, authInit } from '../../../redux/ducks/auth';
import { Link } from 'react-router-dom';

import Page from '../../common/Page';
import buttonClickSound from '../../../assets/soundFX/buttonClick.mp3';

function SignUpPage({ authenticateSignUp, error, authInit }) {
  const [username, setUsername] = useState("");
  const [email, setEmail] = useState("");
  const [passwordOne, setPasswordOne] = useState("");
  const [passwordTwo, setPasswordTwo] = useState("");

  useEffect(() => {
    authInit()
  }, [authInit])

  const [playMouseClickSound] = useSound(
    buttonClickSound,
    {
      volume: 0.5
    }
  )

  const onSubmit = e => {
    e.preventDefault();
    playMouseClickSound();
    authenticateSignUp({ email, username, passwordOne, passwordTwo })
  }

  return (
    <Page isMain={true} isNotLoggedIn={true}>
      <div className="auth-container">
        <h1>Sign Up</h1>
        <form className="auth-form" onSubmit={onSubmit}>
          <div className="fill-in-form">
            <div className="fill-in">
              <div className="fill-name-container">
                <h2 className="fill-name">Username</h2>
              </div>
              <div className="input-container">
                <input
                  data-testid="usernameField"
                  name="username"
                  className="input"
                  value={username}
                  onChange={e => setUsername(e.target.value)}
                  type="text"
                  placeholder="Username"
                />
              </div>
            </div>
            <div className="fill-in">
              <div className="fill-name-container">
                <h2 className="fill-name">Email</h2>
              </div>
              <div className="input-container">
                <input
                  data-testid="emailField"
                  className="input"
                  name="email"
                  value={email}
                  onChange={e => setEmail(e.target.value)}
                  type="email"
                  placeholder="Email Address"
                />
              </div>
            </div>
            <div className="fill-in">
              <div className="fill-name-container">
                <h2 className="fill-name">Password</h2>
              </div>
              <div className="input-container">
                <input
                  data-testid="passwordOneField"
                  className="input"
                  name="passwordOne"
                  value={passwordOne}
                  onChange={e => setPasswordOne(e.target.value)}
                  type="password"
                  placeholder="Password"
                  // autoComplete="new-password"
                  minLength="6"
                />
              </div>
            </div>
            <div className="fill-in">
              <div className="fill-name-container">
                <h2 className="fill-name">Confirm Password</h2>
              </div>
              <div className="input-container">
                <input
                  data-testid="passwordTwoField"
                  className="input"
                  name="passwordTwo"
                  value={passwordTwo}
                  onChange={e => setPasswordTwo(e.target.value)}
                  type="password"
                  placeholder="Confirm Password"
                  // autoComplete="new-password"
                  minLength="6"
                />
              </div>
            </div>
            <div className="menu-buttons">
              {error &&
                <div >
                  <p data-testid="auth-error-signup" className="auth-error">{error}</p>
                  <br />
                </div>}
            </div>
            <div className="menu-buttons">
              <button data-testid="signUpButton" type="submit" className="btn btn-primary">Sign Up</button>
            </div>
          </div>
        </form>
        <br />
        <Link className="auth-link" to="/signin">Already have an account? Sign in</Link>
      </div>
    </Page>

  );

}

const mapStateToProps = state => ({
  error: selectAuthError(state),
});

const dispatchers = {
  authenticateSignUp,
  authInit
};

export default connect(mapStateToProps, dispatchers)(SignUpPage);