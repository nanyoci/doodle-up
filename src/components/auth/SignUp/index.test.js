import React from 'react'
import { cleanup, fireEvent, waitForElement, act } from '@testing-library/react'
import axiosMock from 'axios';

import SignUpPage from './index.js';
import { renderWithReduxRouter } from './../../../utils/tests';
import { errorAction } from './../../../redux/ducks/auth';

jest.mock('axios');

beforeEach(() => {
    window.HTMLMediaElement.prototype.load = () => { /* do nothing */ };
    window.HTMLMediaElement.prototype.play = () => { /* do nothing */ };
    window.HTMLMediaElement.prototype.pause = () => { /* do nothing */ };
    window.HTMLMediaElement.prototype.addTextTrack = () => { /* do nothing */ };
    localStorage.removeItem('username');
})

afterEach(() => {
    cleanup();
    jest.clearAllMocks();
});

it('should take a snapshot', async () => {
    let asFragment = {};
    await act(async () => { ({ asFragment } = renderWithReduxRouter(<SignUpPage />)) });
    expect(asFragment()).toMatchSnapshot();
})


it('allows the user to register successfully', async () => {
    const dummyResponse = "User created."
    const username = "DoodleUp"

    axiosMock.post.mockResolvedValueOnce({
        "data": dummyResponse
    });

    const { getByTestId } = renderWithReduxRouter(<SignUpPage />);
    const emailField = await waitForElement(() => getByTestId("emailField"));
    const usernameField = await waitForElement(() => getByTestId("usernameField"));
    const passwordOneField = await waitForElement(() => getByTestId("passwordOneField"));
    const passwordTwoField = await waitForElement(() => getByTestId("passwordTwoField"));
    const registerButton = await waitForElement(() => getByTestId("signUpButton"));

    await act(async () => {
        fireEvent.change(emailField, {
            target: { value: 'doodleup@test.com' },
        })
        fireEvent.change(usernameField, {
            target: { value: "DoodleUp" },
        })
        fireEvent.change(passwordOneField, {
            target: { value: 'doodleup1234' },
        })
        fireEvent.change(passwordTwoField, {
            target: { value: 'doodleup1234' },
        })
    });

    await act(async () => {
        fireEvent.click(registerButton);
    });

    expect(axiosMock.post).toHaveBeenCalledTimes(1)
    expect(localStorage.getItem('username')).toEqual(username);

})

it('reject user register when passwords do not matched', async () => {
    const username = "DoodleUp"

    const { getByTestId } = renderWithReduxRouter(<SignUpPage />);
    const emailField = await waitForElement(() => getByTestId("emailField"));
    const usernameField = await waitForElement(() => getByTestId("usernameField"));
    const passwordOneField = await waitForElement(() => getByTestId("passwordOneField"));
    const passwordTwoField = await waitForElement(() => getByTestId("passwordTwoField"));
    const registerButton = await waitForElement(() => getByTestId("signUpButton"));

    await act(async () => {
        fireEvent.change(emailField, {
            target: { value: 'doodleup@test.com' },
        })
        fireEvent.change(usernameField, {
            target: { value: username },
        })
        fireEvent.change(passwordOneField, {
            target: { value: 'doodleup1234' },
        })
        fireEvent.change(passwordTwoField, {
            target: { value: ' mismatched passwods' },
        })
    });

    await act(async () => {
        fireEvent.click(registerButton);
    });

    // const err = await waitForElement(() => getByTestId("auth-error-signup"));
    // expect(err.textContent).toEqual("Passwords do not matched.")
    expect(axiosMock.post).toHaveBeenCalledTimes(0)
    expect(localStorage.getItem('username')).toEqual(null)
})


it('reject user register when email is taken', async () => {
    const dummyResponse = "The email is already in use"
    const username = "DoodleUp"

    axiosMock.post.mockResolvedValueOnce({
        "data": dummyResponse
    });

    // errorAction.mockResolvedValueOnce()

    const { getByTestId } = renderWithReduxRouter(<SignUpPage />);
    const emailField = await waitForElement(() => getByTestId("emailField"));
    const usernameField = await waitForElement(() => getByTestId("usernameField"));
    const passwordOneField = await waitForElement(() => getByTestId("passwordOneField"));
    const passwordTwoField = await waitForElement(() => getByTestId("passwordTwoField"));
    const registerButton = await waitForElement(() => getByTestId("signUpButton"));
    await act(async () => {


        fireEvent.change(emailField, {
            target: { value: 'doodleup@test.com' },
        })
        fireEvent.change(usernameField, {
            target: { value: username },
        })
        fireEvent.change(passwordOneField, {
            target: { value: 'doodleup1234' },
        })
        fireEvent.change(passwordTwoField, {
            target: { value: 'doodleup1234' },
        })
    });
    await act(async () => {
        fireEvent.click(registerButton)
    });

    // const err = await waitForElement(() => getByTestId("auth-error-signup"));
    // expect(err).toBeVisible();
    // console.log(err)
    // expect(err).toEqual("Oops! There is a problem with registration.")
    // expect(errorAction).toHaveBeenCalledTimes(1)
    // expect(errorAction).toHaveBeenCalledWith('Oops! There is a problem with registration.')
    expect(axiosMock.post).toHaveBeenCalledTimes(1)
    expect(localStorage.getItem('username')).toEqual(null);
})

