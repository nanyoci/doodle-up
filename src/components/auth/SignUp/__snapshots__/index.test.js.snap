// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="main-content"
  >
    <audio />
    <div
      class="wrapper"
    >
      <img
        alt="DoodleUp"
        class="logo-center"
        src="doodle-up.svg"
      />
      <img
        alt="Open book"
        class="book-corner"
        src="book.svg"
      />
      <img
        alt="Blue splat of paint"
        class="splat-blue-corner"
        src="splat-blue.png"
      />
      <img
        alt="Pink splat of paint"
        class="splat-pink-corner"
        src="splat-pink.png"
      />
      <img
        alt="Paint brush and palette"
        class="paint-corner"
        src="paint.svg"
      />
      <div
        class="auth-container"
      >
        <h1>
          Sign Up
        </h1>
        <form
          class="auth-form"
        >
          <div
            class="fill-in-form"
          >
            <div
              class="fill-in"
            >
              <div
                class="fill-name-container"
              >
                <h2
                  class="fill-name"
                >
                  Username
                </h2>
              </div>
              <div
                class="input-container"
              >
                <input
                  class="input"
                  data-testid="usernameField"
                  name="username"
                  placeholder="Username"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="fill-in"
            >
              <div
                class="fill-name-container"
              >
                <h2
                  class="fill-name"
                >
                  Email
                </h2>
              </div>
              <div
                class="input-container"
              >
                <input
                  class="input"
                  data-testid="emailField"
                  name="email"
                  placeholder="Email Address"
                  type="email"
                  value=""
                />
              </div>
            </div>
            <div
              class="fill-in"
            >
              <div
                class="fill-name-container"
              >
                <h2
                  class="fill-name"
                >
                  Password
                </h2>
              </div>
              <div
                class="input-container"
              >
                <input
                  class="input"
                  data-testid="passwordOneField"
                  minlength="6"
                  name="passwordOne"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              class="fill-in"
            >
              <div
                class="fill-name-container"
              >
                <h2
                  class="fill-name"
                >
                  Confirm Password
                </h2>
              </div>
              <div
                class="input-container"
              >
                <input
                  class="input"
                  data-testid="passwordTwoField"
                  minlength="6"
                  name="passwordTwo"
                  placeholder="Confirm Password"
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              class="menu-buttons"
            />
            <div
              class="menu-buttons"
            >
              <button
                class="btn btn-primary"
                data-testid="signUpButton"
                type="submit"
              >
                Sign Up
              </button>
            </div>
          </div>
        </form>
        <br />
        <a
          class="auth-link"
          href="/signin"
        >
          Already have an account? Sign in
        </a>
      </div>
    </div>
    <div
      class="control-buttons"
    >
      <button
        class="ui yellow massive circular icon button"
        id="musicControlBtn"
      >
        <i
          aria-hidden="true"
          class="mute icon"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
